src/embeddings/tools/sgns.cpp

```diff
@@ -0,0 +1,573 @@
+/**
+ * @file sgns.cpp
+ * @author Shawn LeMaster
+ * @author Matt Kelly
+ *
+ * Builds word embedding vectors using the Skip-Gram Negative Sampling method.
+ *
+ * @see https://code.google.com/archive/p/word2vec/
+ *
+ * All files in META are dual-licensed under the MIT and NCSA licenses. For more
+ * details, consult the file LICENSE.mit and LICENSE.ncsa in the root of the
+ * project.
+ */
+
+#include <atomic>
+#include <chrono>
+#include <deque>
+#include <iostream>
+#include <memory>
+#include <random>
+
+#include "cpptoml.h"
+#include "meta/analyzers/all.h"
+#include "meta/analyzers/token_stream.h"
+#include "meta/corpus/corpus_factory.h"
+#include "meta/hashing/probe_map.h"
+#include "meta/io/filesystem.h"
+#include "meta/io/packed.h"
+#include "meta/logging/logger.h"
+#include "meta/parallel/thread_pool.h"
+#include "meta/util/aligned_allocator.h"
+#include "meta/util/printing.h"
+#include "meta/util/progress.h"
+
+using namespace meta;
+
+class sgns_local_buffer
+{
+    public:
+        sgns_local_buffer(const analyzers::token_stream& stream)
+            : stream_{stream.clone()}
+        { }
+
+        sgns_local_buffer(sgns_local_buffer&&) = default;
+
+    private:
+        friend class sgns_trainer;
+        std::unique_ptr<analyzers::token_stream> stream_;
+};
```

Comment:
The issue here was that the parallel_consume passes this into a function which causes it to attempt to copy the unqiue_ptr causing a compile error. That said, I think we'll move some things into it as mentioned below.